File: 4 Kyu/4 Kyu - Getting along with Integer Partitions.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

struct stat {
    int range;
    double mean;
    double median;
};
int cmpfunc (const void * a, const void * b) {
   return ( *(int*)a - *(int*)b );
}
struct stat compute (int *arr, int size) {

    struct stat actual = {0};
    int next = 0, data[size];

    qsort (arr,size,sizeof(int),cmpfunc);

    for (int i = 0; i < size; ++i)
        if (arr[i] != arr[i+1])
            actual.mean += data[next++] = arr[i];

    actual.median = (next % 2 == 0) ? ((data[next/2] + data[next/2 - 1]) / 2.0) : data[next/2];

    actual.range = data[next - 1] - 1;
    actual.mean /= (float)next;

    return actual;
}
int mult (int *arr, int size) {

    int prod = 1;
    for (int i = 0; i < size; ++i)
        prod *= arr[i];

    return prod;
}
char* part(int n) {

    int mem = 64, size = 0, *prods = malloc (mem * sizeof(int));
    char *output = malloc (64 * sizeof (char));
    int next = 0, stack, digit, index = 0, memstk = 0, part[n];
    struct stat actual;

    memset (part, 0, sizeof(part));
    part[index] = n + 1;

    while (part[0] > 1) {

        digit = part[index] - 1;
        stack = memstk, next = index;

        while (stack < n)
            if (n - stack >= digit) {
                if (digit > 1) {
                    index = next, memstk = stack;
                }
                stack += part[next++] = digit;

            }
            else
                digit--;

        prods[size++] = mult (part,next);

        if (size == mem) {
            mem += mem;
            prods = realloc (prods, mem * sizeof (int));
        }

        if (part[index] == 1) index--, memstk -= part[index];
        //if (part[0] == 1) break;
    }

    actual = compute (prods,size);
    sprintf (output,"Range: %i Average: %.2f Median: %.02f",actual.range,actual.mean,actual.median);

    return output;
}

int main () {

    part (4);
}

File: 4 Kyu/4 Kyu - Getting along with Integer Partitions.rs
use std::collections::BTreeSet;

fn part (num:i64) -> String {

    let mut cluster:Vec<i64> = vec![0;num as usize];
    let mut set:BTreeSet<i64> = BTreeSet::new();

    let mut index:usize = 0;
    let mut prev:i64 = 0;

    cluster[index] = num + 1;

    while cluster[0] > 1 {

        let mut digit = cluster[index] - 1;
        let mut next = index;
        let mut rem = prev;

        while rem < num {
            if num - rem >= digit {
                if digit > 1 {
                    index = next;
                    prev = rem;
                }

                cluster[next] = digit;
                rem += digit;
                next += 1;
            } else {
                digit -= 1;
            }
        }

        set.insert (cluster.iter().filter(|&&x| x != 0).product::<i64>());

        if index > 0 && cluster[index] == 1 {
            index -= 1;
            prev -= cluster[index];
        }
    }

    let hist =  set.iter().map(|&x| x as f64).collect::<Vec<f64>>();
    let size = hist.len();

    let mean:f64 = hist.iter().sum::<f64>() / size as f64;
    let range = hist[size - 1] - hist[0];
    let median:f64;

    if size % 2 == 0 { median = (hist[size / 2] + hist[size / 2 - 1]) / 2.0 } else { median = hist[size / 2] }

    format! ("Range: {range} Average: {mean:.2} Median: {median:.2}")
}
